### ISSUE TITLE
google.iam related unit tests fail with TypeError: No positional arguments allowed

### ISSUE THREAD

**Issue by busunkim96** [googleapis/gapic-generator-python#314]
Possibly related to #205 

Steps to replicate:
1. Generate [billing v1](https://github.com/googleapis/googleapis/tree/master/google/cloud/billing/v1) with the latest Docker image published.
2. Run unit tests.
```
nox -s unit-3.6
```

The tests failures are all involve `google.iam`.  The setup.py didn't have [`grpc-google-iam-v1`](https://pypi.org/project/grpc-google-iam-v1/), so I added it manually (should I be using a different package?)

```py
from google.iam.v1 import iam_policy_pb2 as iam_policy  # type: ignore
from google.iam.v1 import policy_pb2 as policy  # type: ignore
```

```python
tests/unit/billing_v1/test_cloud_billing.py:938: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <google.cloud.billing_v1.services.cloud_billing.client.CloudBillingClient object at 0x7f095f303bd0>, request = None

    def test_iam_permissions(
        self,
        request: iam_policy.TestIamPermissionsRequest = None,
        *,
        resource: str = None,
        permissions: Sequence[str] = None,
        retry: retries.Retry = gapic_v1.method.DEFAULT,
        timeout: float = None,
        metadata: Sequence[Tuple[str, str]] = (),
    ) -> iam_policy.TestIamPermissionsResponse:
        r"""Tests the access control policy for a billing
        account. This method takes the resource and a set of
        permissions as input and returns the subset of the input
        permissions that the caller is allowed for that
        resource.
    
        Args:
            request (:class:`~.iam_policy.TestIamPermissionsRequest`):
                The request object. Request message for
                `TestIamPermissions` method.
            resource (:class:`str`):
                REQUIRED: The resource for which the policy detail is
                being requested. ``resource`` is usually specified as a
                path. For example, a Project resource is specified as
                ``projects/{project}``.
                This corresponds to the ``resource`` field
                on the ``request`` instance; if ``request`` is provided, this
                should not be set.
            permissions (:class:`Sequence[str]`):
                The set of permissions to check for the ``resource``.
                Permissions with wildcards (such as '*' or 'storage.*')
                are not allowed. For more information see `IAM
                Overview <https://cloud.google.com/iam/docs/overview#permissions>`__.
                This corresponds to the ``permissions`` field
                on the ``request`` instance; if ``request`` is provided, this
                should not be set.
    
            retry (google.api_core.retry.Retry): Designation of what errors, if any,
                should be retried.
            timeout (float): The timeout for this request.
            metadata (Sequence[Tuple[str, str]]): Strings which should be
                sent along with the request as metadata.
    
        Returns:
            ~.iam_policy.TestIamPermissionsResponse:
                Response message for ``TestIamPermissions`` method.
        """
        # Create or coerce a protobuf request object.
        # Sanity check: If we got a request object, we should *not* have
        # gotten any keyword arguments that map to the request.
        if request is not None and any([resource, permissions]):
            raise ValueError(
                "If the `request` argument is set, then none of "
                "the individual field arguments should be set."
            )
    
        # If we have keyword arguments corresponding to fields on the
        # request, apply these.
>       request = iam_policy.TestIamPermissionsRequest(request)
E       TypeError: No positional arguments allowed

google/cloud/billing_v1/services/cloud_billing/client.py:966: TypeError
```

**Comment by software-dov** [comment]
Just reproduced the issue. Yep, that looks related.

**Comment by software-dov** [comment]
All that's required to reproduce this is a method whose request type definition lives in a proto not passed to the generator, i.e. it's just imported into the proto where the method is defined.
That means that the type imported and used for the wrapping isn't a [proto-plus](https://proto-plus-python.readthedocs.io/) wrapped class, just a plain, vanilla python protobuf wrapper, which has a much less ergonomic API.

There are a few possible solutions:
1. If the request type is external, add a check in the generated code if the input parameter is a dict. If it is, double-star expand it into the request type. If it isn't, use the passed parameter directly.
2. Add some complex protobuf introspection logic.
3. Require that all protobuf message types be proto-plus wrappers created by the generator.

I think that 1 is flexible enough to seamlessly allow wrapped types if they are passed, and the expansion/detection logic shouldn't be too egregious. I don't want to do 2, and 3 seems too inflexible in the greater ecosystem.

I'll open a separate issue to track the missing import, but I'm not sure if that can be handled automatically because of the discoverability of the package name.

**Comment by software-dov** [comment]
The requirement issue in setup.py is tracked by #316

**Comment by busunkim96** [comment]
@software-dov Would publishing a new version of `grpc-google-iam-v1` iam that uses the proto-plus classes be an option?

**Comment by software-dov** [comment]
That would definitely solve the immediate issue, but I'd also like to have a fix in place for the general case. We can deprioritize patching the generator if you think that iam is special and that the general case is sufficiently uncommon to not be an issue.

**Comment by busunkim96** [comment]
I think iam is a special snowflake, and we generally discourage people from writing APIs that use protos from other APIs.

IRM [depends on monitoring](https://github.com/googleapis/python-irm/blob/master/setup.py#L28) and I think there is one more API (Asset?) that will do it in the near term.

Any idea how grpc-google-iam-v1 is published?

**Comment by software-dov** [comment]
No clue. It's maintained by https://pypi.org/user/googleapis-packages/, but that doesn't really tell us much about the generation process.

**Comment by software-dov** [comment]
Fixed in-generator by #317 , just verified that Billing generates and passes unit tests.
